virtual_ecosystem/core/variables.py

```diff
@@ -0,0 +1,290 @@
+"""Module for all variables."""
+
+import inspect
+import pkgutil
+from collections.abc import Hashable
+from dataclasses import asdict, dataclass, field
+from importlib import import_module
+from pathlib import Path
+from typing import cast
+
+import tomli_w
+
+import virtual_ecosystem.core.axes as axes
+import virtual_ecosystem.core.base_model as base_model
+from virtual_ecosystem.core.logger import LOGGER
+
+
+@dataclass
+class Variable:
+    """Simulation variable, containing static and runtime metadata."""
+
+    name: str
+    """Name of the variable. Must be unique."""
+    model_name: str
+    """Name of the model defining the variable."""
+    description: str
+    """Description of what the variable represents."""
+    units: str
+    """Unites the variable should be represented in."""
+    var_type: str
+    """Type of the variable."""
+    axis: tuple[str, ...]
+    """Axes the variable is defined on."""
+    initialised_by: str = field(default_factory=str, init=False)
+    """Model that initialised the variable."""
+    required_init_by: list[str] = field(default_factory=list, init=False)
+    """Models that requires the variable to be initialised."""
+    updated_by: list[str] = field(default_factory=list, init=False)
+    """Models that update the variable."""
+    required_update_by: list[str] = field(default_factory=list, init=False)
+    """Models that use the variable."""
+
+    def __post_init__(self) -> None:
+        """Register the variable in the known variables.
+
+        Raises:
+            ValueError: If a variable is already in the known variables registry.
+        """
+        if self.name in KNOWN_VARIABLES:
+            raise ValueError(
+                f"Variable {self.name} already in the known variables registry."
+            )
+
+        LOGGER.info(f"Variable registered for model '{self.model_name}': {self.name}")
+        KNOWN_VARIABLES[self.name] = self
+
+
+RUN_VARIABLES_REGISTRY: dict[str, Variable] = {}
+"""The global registry of variables used in a run."""
+
+KNOWN_VARIABLES: dict[str, Variable] = {}
+"""The global known variable registry."""
+
+
+def register_all_variables() -> None:
+    """Registers all variables provided by the models."""
+    import virtual_ecosystem.models as models
+
+    for mod in pkgutil.iter_modules(models.__path__):
+        if not mod.ispkg:
+            continue
+
+        try:
+            import_module(f"{models.__name__}.{mod.name}.variables")
+        except ImportError:
+            LOGGER.debug(
+                f"No 'variables' module found for model {models.__name__}.{mod.name}."
+            )
```

Comment:
Actually - looking at the way `ve_run`  works at the moment, it loops over _all_ defined models, whether they are used or not in a run. That:

* sort of doesn't matter, because all we're doing is populating a registry with more information than is needed,  
* means we can share the same function between `ve_run` and `output_known_variables`

But it does also mean that logs will contain messages about models that aren't being used. I think it might make more sense to include lines 74-78 in `core.registry.register_module`. That gets run in the next line in `ve_run` - as each model in the config files is added to the configuration, the `MODULE_REGISTRY` gets populated with `ModuleInfo` instances. We don't want to add information by module, but the same function could be used to populate the separate cross module `RUN_VARIABLES_REGISTRY`?

The configuration build is explicitly used to only register the modules included in the config, so that variable registry then only contains the supposed set of required variables, which is probably better for spotting bad model setups or missing data?

Somewhere after this - finish adding the `ModuleInfo` and then populate the variables?

https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/89169fbb1e3e1e7cdb5d963c0bd141bc122f7df3/virtual_ecosystem/core/registry.py#L156

I think `_discover_all_variables_usage` and `register_all_variables` probably then collapse into `output_known_variables`, although we could keep it modular?

Reply:
I think there is value on keeping things modular and concise, doing only one thing. Having said that, I'll check the best way of reducing redundancy when registering stuff. You suggestion seems sensible.

Reply:
I'm checking your suggestions in detail and I think that only the requested modules are registered, and not all of the available ones. The `requested modules` scanned in the `Config` and used to build the registry are only those included in the configuration files, in turn passed as `config_paths` or `config_strings`, which are provided by the user. Am I missing something?

Reply:
Yes, but `register_all_variables` scans _all_ of the `model.variable` modules. We already have duplicate parallel models - and its something we want to encourage - so those `variables.py` are going to contain duplicates and not all are needed for the specific run, so loading all modules seem unnecessary and possible a problem.

I think there's a bigger question  - I haven't quite got my head around _where_ we put variables. The `models.variables` module seems like a natural place, but where do we put variables always loaded from data? They don't really belong in any one model - some may only be used by one model but others will overlap, so how do model designers know which to include where. The same issue occurs with parallel models - we have repetition across files and multiple places to edit if we change definitions.

It seems more DRY and easy to manage if we just have one variable data source - an integrated home outside the models where all definitions get loaded from. That could even be the existing `data_variables.toml` file, but stripped of the model usage, which this system can then populate for the configured models.  That will break when we come to consider how users can plug in models - we probably want to be able to specify a model directory to add local models, but that could provide another `data_variables.toml` source file that _must_ not repeat.

Reply:
I'm not sure loading all known variables is a problem - it's just a lookup table for possible values - but the overlap, duplication and uncertainty about which module a variable should go in seems undesireable.

Reply:
I'm not sure about variables loaded form data, to be honest, but model variables should live in their own module within the appropriate folder. For parallel models, designers just need to check if all the variables they need have already been defined in another model of the same kind - or anywhere else - and define only those missing. That's were the documentation comes into play.

Reply:
I think there's a case for defining variables in one location - a root variables module maybe? I haven't entirely thought this through but it seems wrong to have parallel models having to contain the same variable definitions. Each model might have slightly different sets of variables, but it doesn't seem DRY to have that duplication.

Reply:
As far as I can tell, there are two options:
- A root `variables.py` module that contains ALL variables and where new variables are added. Easy to maintain and to monitor for changes. Will make the code simpler and solves the issue of where to put the input data variables. I don't see a major drawback except, maybe, that you have variables for all models all mixed up together, which somewhat reduces modularity (a model is no longer a self-contained directory).
- A _per model_ `variables.py` module, BUT not duplicating variables, i.e. if the `adiabatic` model already defines variables A, B and C, `adiabatic_simple` does not need to defined them again - indeed, it would result in an error, as the code is now. This groups variables that are conceptually related together, but might make it harder to trace where they are actually defined, and doesn't solve the issue of input data variables. As variable definition is not duplicated, modularity is also not that great.

Duplicating the variable definition I don't think is a solution, so I would not go down that route even if it makes the code more modular and self'contained.

My preference would be the first option.

Reply:
I've just realised that my last message repeats stuff I've already said...

Reply:
Ok, I'll review the code based on this new interpretation this week (possibly tomorrow or on Thursday). I won't enter into the "providing external models" idea as that might require further thought, given that it might have more ramifications. 

@vgro don't add any variable, yet, as the implementation is still fluid.

Reply:
@dalonsoa The external models is definitely for another PR. I just wanted to flag that we will definitely need to do this, but I don't think the implementation will be super hard if we take the new route.

Reply:
I also agree that option 1 seems the cleanest. And while it does make the system slightly less modular, I kind of doubt that the main limit on modularity will come from the code side anyway (I think it'll mainly come from the theory/science/concept definition side)

Reply:
@dalonsoa Just to add that we all chatted about this in a catch up today and we all agree that Option 1 looks good.